hashintel/hash
# Support link arrays on links table at entity editor page

## 🌟 What is the purpose of this PR?

This PR expands the functionality of the **links table** on the **edit entity** page. With these changes, users can link and unlink entities on link lists (links where the value of `maxItems` is greater than `1`). This PR also changes the links table design based on the current status of (ongoing) internal discussions.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/0/1203199644802795) _(internal)_
- [Design](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH-App?node-id=6468%3A253538&t=5AeJC1C6PDLgoPPZ-4) _(internal)_
  - ⚠️ The implemented designs deviate from the originals in a number of places. Follow-up decisions need to be made to adjust either the designs or the code. Follow-up tickets will be created.
- [Slack thread](https://hashintel.slack.com/archives/C022217GAHF/p1670414417826489) _(internal)_ where we discuss some of the changes

## 🔍 What does this change?

- Add optional `opacity` support to `drawCellFadeOutGradient` function
- Design changes on links table
- Make cursor `pointer` on `linkedWith` column
- Add `title` prop to `AddAnotherButton` component
- Add `icon` & `tooltip` props to `ValueChip` component

## ⚠️ Known issues

There are some known issues, most of them being UX related. We need to discuss how to solve those issues efficiently without wasting effort. Because we'll eventually implement two big changes on top of entity editor.

Those big changes are `collab support` and `draft entity state`. These two can potentially change the entity-editor's state management drastically, so it could be clever to avoid spending more effort on fixing UX issues on the links table if those UX issues will be solved by the changes we'll introduce with `collab` & `draft entity state`.

### Issues of links table
Removing / adding links take some time. And we’re not doing actual optimistic updates, because it’s hard to update the entity subgraph manually for links. So instead of optimistic updates, we just `refetch` the entity subgraph on changes. That causes some UX issues. 

### Technical explanation of those UX issues
If users link/unlink entities in quick succession, race conditions happen. Since we didn’t implemented any loading state for the table due to avoid wasting effort. Because eventually table will update immediately without a need for loading state. Eventually we’ll d (A) optimistic update or (B) draft entity state managed on client-side.

We decided to open this PR with it’s current state with @CiaranMn, because it has the actual functionality working, even if with some UX flaws. If users does not try to link/unlink entities in a quick succession, it should be fine.

### Some demos of the UX issues / bugs happen when users link/unlink entities in quick succession

https://user-images.githubusercontent.com/39553853/206283861-8b2eef2b-6651-4b46-aec0-926b4b386e58.mp4



## 🐾 Next steps

- Fix the known issues by adressing them individually, or wait for `collab` & `draft entity state` changes to see how the code is going to change around these issues.

## ❓ How to test this?


1.  Checkout the branch / view the deployment
1.  Open up entity editor for any entity, and use the [create demo entity type button](https://github.com/hashintel/hash/commit/202d6c0103a89e13bd8e75a9951cbb22945b29ee) to create a demo entity type
1.  Confirm that links table works as shown in the demo below.

## 📹 Demo

https://user-images.githubusercontent.com/39553853/206265294-3157c259-3dbb-45a2-b7a6-1108e8db68a4.mov
